### DISCUSSION TITLE
3.0开发建议

### DISCUSSION THREAD

**Discussion by s96a5g9f8iyh** [layui/layui#1575] [建议]
前段时间，发现了一个alpine.js的轻型 框架

vue等等框架流行，有部分提高生产力的原因。
例如生命周期 对于逻辑开发降低了心智负担。
例如双向绑定 对于下拉框这种列表的生成 很方便，而如果用jq生成的话 就需要更多的代码。

现在3.0不是在开发中吗，可以看看我推荐的这个轻型框架。参考 例如双向绑定 等等
下方示例 是我学习时候 写的demo

`

<head>

    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.3/dist/cdn.min.js"></script>
    

</head>

<body>
    
    <div x-data="demo">
        
        <p x-text="message"></p>

        
        <div x-html="htmlContent"></div>

        
        <div x-show="show">
            显示隐藏
        </div>

        
        <img :src="demosrc" alt="">

        
        <p x-text="add"></p>

        
        <input type="text" x-model="message" placeholder="在这里输入消息">
        
        <p>输入框内容： <span x-text="message"></span></p>

        
        <button @click="showMessage" x-text="name"></button>

        
        <ul>
            
            <template x-for="item in items">
                <li>
                    <span x-text="item"></span>
                    <button @click="showTabs(`${item}`)">传递参数</button>
                </li>
            </template>
        </ul>
        
        <button @click="items.push('Pear')">添加水果</button>

        
        <ul x-data="{ car: { make: 'Jeep', model: 'Grand Cherokee', color: 'Black' } }">
            <template x-for="(value, index) in car">
                <li>
                    <span x-text="index"></span>: <span x-text="value"></span>
                </li>
            </template>
        </ul>

        
        <ul x-data="{ colors: [
            { id: 1, label: 'Red' },
            { id: 2, label: 'Orange' },
            { id: 3, label: 'Yellow' },
        ]}">
            <template x-for="color in colors" :key="color.id">
                <li x-text="color.label"></li>
            </template>
        </ul>

        
        <button @click="show = !show">切换登录状态</button>
        <template x-if="show">
            <p>已登录！</p>
            <button @click="show = false">退出登录</button>
        </template>
        <template x-if="!show">
            <p>未登录！</p>
            <button @click="show = true">登录</button>
        </template>

        
        <button @click="$refs.text.remove()">Remove Text</button>
        
        <span x-ref="text">Hello 👋</span>

        <a href="./index1.html">跳转</a>

        
        <div x-text="$store.darkMode.on"></div>
        <button x-data @click="$store.darkMode.toggle()">Toggle Dark Mode</button>

    </div>

    
    <div x-data="demo1">
        <div x-text="$store.darkMode.on"></div>

        
        <div @foo="console.log('foo was dispatched')">
            
            <button @click="$dispatch('foo')">自定义事件</button>
        </div>

        
        <div @notify="alert($event.detail.message)">
            
            <button @click="$dispatch('notify', { message: 'Hello World!' })">
                自定义事件传递参数
            </button>
        </div>
    </div>

    
    <div x-data="{ title: 'Hello' }" @set-title.window="title = $event.detail">
        <h1 x-text="title"></h1>
    </div>
    
    <div x-data>
        <button @click="$dispatch('set-title', 'Hello World!')">Click me</button>
    </div>

    
    <div x-data="{count:0}">
        <p x-text="count"></p>
        <button @click="count++">增加</button>
        <button @click="count--">减少</button>
    </div>
    
</body>

<script>
    // 文档加载完成后注册事件监听器
    document.addEventListener('alpine:init', () => {
        //注册组件
        Alpine.data(
            // 组件名
            'demo', () => ({
                // 组件属性
                name: "测试按钮",
                message: 'Hello from Global Component!',
                htmlContent: '<h2>Alpine.js x-html 示例</h2><p>这是一段包含 HTML 标记的内容。</p>',
                show: true,
                items: ['Apple', 'Banana', 'Orange'],
                car: { make: 'Jeep', model: 'Grand Cherokee', color: 'Black' },
                count1: 1,
                count2: 2,
                demosrc: "https://tdesign.gtimg.com/site/baseLogo-light.png",
                // 组件计算(getter)
                get add() { return this.count1 + this.count2 },
                // 自定义方法
                showMessage() {
                    alert(this.message);
                },
                showTabs(data_id) {
                    console.log(data_id);
                },
                async getcaptcha() {
                    //请求网址
                    let url = "http://localhost/digit"

                    //构建表单
                    let form = new FormData();
                    form.append('page', this.currentPage);
                    form.append('pagesize', this.pagesize);

                    //请求参数设置（构建表单时，不能添加header头，不然gin会接收不到参数。）
                    let options = {
                        method: "post",
                        // mode: "cors",
                        headers: {
                            // 'Content-Type': 'application/json',
                            // 'Content-Type': 'application/x-www-form-urlencoded',
                            // 'Content-Type': 'text/html;charset=utf-8',
                        },
                        body: form,
                    }

                    //错误处理
                    try {
                        //网络请求
                        let response = await fetch(url, options);

                        if (response.status != 200) {
                            //请求失败 抛出异常
                            throw new Error(response.statusText);
                        }

                        //如果请求成功 返回结果
                        let data = await response.json();

                        //给数据赋值
                        this.id = data.id
                        this.base64 = data.b64s
                        this.answer = data.answer

                        console.log('数据是', data)

                        //捕获异常
                    } catch (error) {
                        //处理异常
                        console.log('Request Failed', error);
                        //也可以不处理异常 将异常继续向上一级抛出
                        // throw error;

                        //最后执行代码
                    } finally {
                        //finally并不是处理异常 而是不管有没有异常都会最后执行的代码
                        console.log('最后执行代码');
                    }
                },
                //初始化
                init() {
                    console.log('I will get evaluated when initializing each "dropdown" component.')
                    // 监听器
                    this.$watch('show', (value, oldValue) => console.log(value, oldValue))
                    // 初始化 调用 自定义方法
                    this.getcaptcha();
                },
                //销毁
                destroy() {
                    console.log('组件销毁')
                    // 销毁处理程序，避免内存泄漏和副作用
                },
            }))

        //注册组件
        Alpine.data(
            // 组件名
            'demo1', () => ({
                message: null,
            }))

        // 跨组件 全局
        Alpine.store('darkMode', {
            on: false,

            toggle() {
                this.on = !this.on
            }
        })
    })
</script>

`

**Comment by s96a5g9f8iyh** [comment]
```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.3/dist/cdn.min.js"></script>
    <!-- <script src="https://cdn.jsdelivr.net/npm/alpinejs@3"></script> -->

</head>

<body>
    <!-- 定义组件 -->
    <div x-data="demo">
        <!-- 动态更新文本 -->
        <p x-text="message"></p>

        <!-- 动态更新html -->
        <div x-html="htmlContent"></div>

        <!-- 显示隐藏 -->
        <div x-show="show">
            显示隐藏
        </div>

        <!-- 绑定属性 -->
        <img :src="demosrc" alt="">

        <!-- 组件计算(getter) -->
        <p x-text="add"></p>

        <!-- 使用 x-model 指令双向绑定输入框的值 -->
        <input type="text" x-model="message" placeholder="在这里输入消息">
        <!-- 显示输入框内容 -->
        <p>输入框内容： <span x-text="message"></span></p>

        <!-- 监听点击时间调用自定义方法 -->
        <button @click="showMessage" x-text="name"></button>

        <!-- 遍历数组 -->
        <ul>
            <!-- 使用 x-for 指令循环渲染数组元素 -->
            <template x-for="item in items">
                <li>
                    <span x-text="item"></span>
                    <button @click="showTabs(`${item}`)">传递参数</button>
                </li>
            </template>
        </ul>
        <!-- 添加新项目到列表 -->
        <button @click="items.push('Pear')">添加水果</button>

        <!-- 遍历map -->
        <ul x-data="{ car: { make: 'Jeep', model: 'Grand Cherokee', color: 'Black' } }">
            <template x-for="(value, index) in car">
                <li>
                    <span x-text="index"></span>: <span x-text="value"></span>
                </li>
            </template>
        </ul>

        <!-- 遍历 并且设置唯一key -->
        <ul x-data="{ colors: [
            { id: 1, label: 'Red' },
            { id: 2, label: 'Orange' },
            { id: 3, label: 'Yellow' },
        ]}">
            <template x-for="color in colors" :key="color.id">
                <li x-text="color.label"></li>
            </template>
        </ul>

        <!-- 使用 x-if 指令进行条件渲染 -->
        <button @click="show = !show">切换登录状态</button>
        <template x-if="show">
            <p>已登录！</p>
            <button @click="show = false">退出登录</button>
        </template>
        <template x-if="!show">
            <p>未登录！</p>
            <button @click="show = true">登录</button>
        </template>

        <!-- 通过魔术属性 获取到 text 然后删除 -->
        <button @click="$refs.text.remove()">Remove Text</button>
        <!-- 设置引用 text -->
        <span x-ref="text">Hello 👋</span>

        <a href="./index1.html">跳转</a>

        <!-- 跨组件 全局 -->
        <div x-text="$store.darkMode.on"></div>
        <button x-data @click="$store.darkMode.toggle()">Toggle Dark Mode</button>

    </div>

    <!-- 定义组件 -->
    <div x-data="demo1">
        <div x-text="$store.darkMode.on"></div>

        <!-- 监听自定义 foo 事件 -->
        <div @foo="console.log('foo was dispatched')">
            <!-- 触发自定义 foo 事件 -->
            <button @click="$dispatch('foo')">自定义事件</button>
        </div>

        <!-- 监听自定义 notify 事件 并接收参数 -->
        <div @notify="alert($event.detail.message)">
            <!-- 触发自定义 notify 事件 并传递参数 -->
            <button @click="$dispatch('notify', { message: 'Hello World!' })">
                自定义事件传递参数
            </button>
        </div>
    </div>

    <!-- 定义组件 监听全局自定义事件-->
    <div x-data="{ title: 'Hello' }" @set-title.window="title = $event.detail">
        <h1 x-text="title"></h1>
    </div>
    <!-- 定义组件 触发全局自定义事件-->
    <div x-data>
        <button @click="$dispatch('set-title', 'Hello World!')">Click me</button>
    </div>

    <!-- 计算器 -->
    <div x-data="{count:0}">
        <p x-text="count"></p>
        <button @click="count++">增加</button>
        <button @click="count--">减少</button>
    </div>
    
</body>
<script>
    // 文档加载完成后注册事件监听器
    document.addEventListener('alpine:init', () => {
        //注册组件
        Alpine.data(
            // 组件名
            'demo', () => ({
                // 组件属性
                name: "测试按钮",
                message: 'Hello from Global Component!',
                htmlContent: '<h2>Alpine.js x-html 示例</h2><p>这是一段包含 HTML 标记的内容。</p>',
                show: true,
                items: ['Apple', 'Banana', 'Orange'],
                car: { make: 'Jeep', model: 'Grand Cherokee', color: 'Black' },
                count1: 1,
                count2: 2,
                demosrc: "https://tdesign.gtimg.com/site/baseLogo-light.png",
                // 组件计算(getter)
                get add() { return this.count1 + this.count2 },
                // 自定义方法
                showMessage() {
                    alert(this.message);
                },
                showTabs(data_id) {
                    console.log(data_id);
                },
                async getcaptcha() {
                    //请求网址
                    let url = "http://localhost/digit"

                    //构建表单
                    let form = new FormData();
                    form.append('page', this.currentPage);
                    form.append('pagesize', this.pagesize);

                    //请求参数设置（构建表单时，不能添加header头，不然gin会接收不到参数。）
                    let options = {
                        method: "post",
                        // mode: "cors",
                        headers: {
                            // 'Content-Type': 'application/json',
                            // 'Content-Type': 'application/x-www-form-urlencoded',
                            // 'Content-Type': 'text/html;charset=utf-8',
                        },
                        body: form,
                    }

                    //错误处理
                    try {
                        //网络请求
                        let response = await fetch(url, options);

                        if (response.status != 200) {
                            //请求失败 抛出异常
                            throw new Error(response.statusText);
                        }

                        //如果请求成功 返回结果
                        let data = await response.json();

                        //给数据赋值
                        this.id = data.id
                        this.base64 = data.b64s
                        this.answer = data.answer

                        console.log('数据是', data)

                        //捕获异常
                    } catch (error) {
                        //处理异常
                        console.log('Request Failed', error);
                        //也可以不处理异常 将异常继续向上一级抛出
                        // throw error;

                        //最后执行代码
                    } finally {
                        //finally并不是处理异常 而是不管有没有异常都会最后执行的代码
                        console.log('最后执行代码');
                    }
                },
                //初始化
                init() {
                    console.log('I will get evaluated when initializing each "dropdown" component.')
                    // 监听器
                    this.$watch('show', (value, oldValue) => console.log(value, oldValue))
                    // 初始化 调用 自定义方法
                    this.getcaptcha();
                },
                //销毁
                destroy() {
                    console.log('组件销毁')
                    // 销毁处理程序，避免内存泄漏和副作用
                },
            }))

        //注册组件
        Alpine.data(
            // 组件名
            'demo1', () => ({
                message: null,
            }))

        // 跨组件 全局
        Alpine.store('darkMode', {
            on: false,

            toggle() {
                this.on = !this.on
            }
        })
    })
</script>

</html>
```

**Reply by lightcax** [reply]
if语句 只允许一个根element one root level element

**Comment by sentsim** [comment]
的确是个很轻量的框架，但 Layui 可能不会去内置它，建议直接在外部使用。
Layui 未来的方向是剔除所有依赖，全面拥抱原生。当然，在此之前，中间仍然会有一个相对更加健全的 jQuery 版本。

**Reply by hulang** [reply]
@sentsim 老大，我们(就代表所有^_^)后端开发者，期待额。。。

**Reply by lightcax** [reply]
@sentsim 老大，3.0 什么时候出来了？

**Reply by KingBes** [reply]
还有就是，我之前写的那些第三方库还能用吗？也跟着重构？
